sources/ghostdog/src/main.rs

```diff
@@ -63,6 +141,44 @@ fn run() -> Result<()> {
             let device_name = find_device_name(format!("{}", path.display()))?;
             emit_device_name(&device_name);
         }
+        SubCommand::MatchNvidiaDriver(vendor) => {
+            // Currently, we only want to run this for NVIDIA devices
+            if vendor.vendor == NVIDIA_VENDOR_ID {
+                let pci_id = env::var("PCI_ID").context(error::MissingPciIdEnvSnafu)?;
+                let driver_choice = find_preferred_driver(pci_id)?;
+                // After choosing the driver, drop a marker file provide this decision to others on the system (like systemd)
+                if !Path::new(NVIDIA_RUN_DIR).exists() {
+                    fs::create_dir_all(NVIDIA_RUN_DIR).context(error::CreateDirSnafu {
+                        path: NVIDIA_RUN_DIR,
+                    })?;
+                }
+                let marker_path: PathBuf = Path::new(NVIDIA_RUN_DIR).join(driver_choice.clone());
+                fs::write(marker_path.clone(), "").context(error::WriteMarkerFileSnafu {
+                    path: marker_path.clone(),
+                })?;
+
+                // Copy the driverdog configuration from the factory into /etc/drivers so the preferred driver is loaded
+                if !Path::new(DEFAULT_DRIVER_CONFIG_PATH).exists() {
+                    fs::create_dir_all(DEFAULT_DRIVER_CONFIG_PATH).context(
+                        error::CreateDirSnafu {
+                            path: DEFAULT_DRIVER_CONFIG_PATH,
+                        },
+                    )?;
+                }
+                let source_driver_path =
+                    Path::new(DRIVER_FACTORY_DIR).join(format!("{}.toml", driver_choice.clone()));
+                let etc_driver_path = Path::new(DEFAULT_DRIVER_CONFIG_PATH)
+                    .join(format!("{}.toml", driver_choice.clone()));
+                fs::copy(source_driver_path.clone(), etc_driver_path.clone()).context(
+                    error::CopyFileSnafu {
+                        s: source_driver_path,
+                        d: etc_driver_path,
+                    },
+                )?;
```

Comment:
This is my major open concern with the change. There's no guarantee that the config file is in place before `driverdog` runs.

You are also making some assumptions (final `/etc` mounted, `/run` mounted) that are true for Bottlerocket but wouldn't hold for many other paths. In general I would avoid these kinds of filesystem operations in a udev context.

Reply:
I've moved away from `udev` so this should avoid these problems.